text/204-golang-bindings.md

```diff
@@ -0,0 +1,762 @@
+}
+
+func (a *Animal) Move(distance int64) {
+    fmt.Printf("%s moved %vm.\n", a.Name(), distance)
+}
+
+// Child class
+type SnakeIface interface {
+  AnimalIface
+}
+
+type Snake struct {
+    Animal
+}
+
+func NewSnake(name string) SnakeIface {
+    return &Snake{Animal{name}}
+}
+
+func (s *Snake) Name() string {
+    return s.name  // inherits `name` property from `Animal`
+}
+
+func (s *Snake) Move(distance int64) {
+    fmt.Printf("Slithering...\n")
+    s.Animal.Move(distance)
+}
+
+// usage:
+dumpling := NewAnimal("Dumpling the Dog")
+dumpling.Move(5)
+
+sam := NewSnake("Sammy the Python")
+sam.Move(10)
+
+// Dumpling the Dog moved 5m.
+// Slithering...
+// Sammy the Python moved 10m.
+```
+
+### Case 3: Class that implements an interface
+
+Using embedding can also be used for a Typescript class that implements an interface:
+
+```ts
+// Typescript
+
+export interface ICluster extends IResource {
+  readonly clusterName: string;
+  readonly clusterArn: string;
+  readonly vpc: ec2.IVpc;
+  readonly connections: ec2.Connections;
+  readonly hasEc2Capacity: boolean;
+  readonly defaultCloudMapNamespace?: cloudmap.INamespace;
+  readonly autoscalingGroup?: autoscaling.IAutoScalingGroup;
+}
+
+export class Cluster extends Resource implements ICluster {
+  public static fromClusterAttributes(scope: Construct, id: string, attrs: ClusterAttributes): ICluster {
+    return new ImportedCluster(scope, id, attrs);
+  }
+  public readonly connections: ec2.Connections = new ec2.Connections();
+  public readonly vpc: ec2.IVpc;
+  public readonly clusterArn: string;
+  public readonly clusterName: string;
+
+  private _defaultCloudMapNamespace?: cloudmap.INamespace;
+  private _hasEc2Capacity: boolean = false;
+  private _autoscalingGroup?: autoscaling.IAutoScalingGroup;
+...
+}
+```
+
+Go struct (derived from TS class) with embedded struct (derived from TS datatype interface):
+
+```go
+type ICluster interface {
+    IResource
+
+    ClusterName()              string
+    ClusterArn()               string
+    Vpc()                      ec2.IVpc
+    Connections()              ec2.Connections
+    HasEc2Capacity()           bool
+    DefaultCloudMapNamespace() cloudmap.INamespace
+    AutoscalingGroup()         autoscaling.IAutoScalingGroup
+}
+
+// Generated interface for Cluster class
+type ClusterIface interface {
+    ICluster
+}
+
+// Generated struct for Cluster class
+type Cluster struct {
+    Resource
+    ICluster
+
+    connections ec2.Connections
+    vpc         ec2.IVpc
+    clusterArn  string
+    clusterName string
+
+    defaultCloudMapNamespace cloudmap.INamespace
+    hasEc2Capacity           bool
+    autoscalingGroup         autoscaling.IAutoScalingGroup
+}
+
+func (c *Cluster) ClusterName() string {
+    return c.clusterName
+}
+
+// ...etc
+```
+
+### Case 4: Abstract Classes
+
+These should be able to be handled much in the same way as regular classes, by generating an interface and a struct. The struct generation is still
```

Comment:
Is there a way to prevent direct instantiation of these structs to mimic the "abstract" part? Maybe a default `New` that throws an error?

Reply:
Hm, perhaps but that wouldn't necessarily stop anyone from trying to declare the struct literal, right?